File: Java/Proyectos/8Puzzle/Solver.java

import java.util.Collections;
import java.util.Comparator;
import java.util.LinkedList;
import java.util.List;

/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

/**
 *
 * @author Usuario UTP
 */
public class Solver {

    private MinPQ<Node> pq;
    private Node goal;
   
    private class Node {
        private Board board;
        private int moves;
        private Node prev;
        
        public Node(Board board) {
            this(board, 0, null);
        }

        public Node(Board board, int moves, Node prev) {
            this.board = board;
            this.moves = moves;
            this.prev  = prev;
        }
        
        public int priority() {
            return board.manhattan() + moves;
        }
        
        @Override
        public String toString() {
            String out = "priority = " + priority();
            out += "\nmoves = " + moves;
            out += "\nmanhattan = " + board.manhattan();
            out += "\n" + board;
            return out;
        }
    }
    
    public Solver(Board board) {
        pq = new MinPQ<Node>(new Comparator<Node>() {
            @Override
            public int compare(Node n1, Node n2) {
                if (n1.priority() > n2.priority()) return 1;
                if (n1.priority() < n2.priority()) return -1;
                return n1.board.manhattan() - n2.board.manhattan();
            }          
        });
        pq.insert(new Node(board));
        
        while (true) {
            Node node = pq.delMin();
            
            if (node.board.isGoal()) {
                goal = node;
                break;
            }
            
            for (Board neighbor : node.board.neighborns()) {
                Node n = new Node(neighbor, node.moves + 1, node);
                if (node.prev != null) {
                    if (!neighbor.equals(node.board)) {
                        pq.insert(n);
                    }
                } else {
                    pq.insert(n);
                }
            }
        }
    }
    
    public int moves() {
        return goal != null ? goal.moves : -1;
    }
    
    public Iterable<Board> soluction() {
        List<Board> sol = new LinkedList<Board>();
        Node node = goal;
        while (node != null) {
            sol.add(node.board);
            node = node.prev;
        }
        Collections.reverse(sol);
        return sol;
    }
    
    public static void main(String[] args) {
        
        int[][] blocks = {{8,1,3},{4,0,2},{7,6,5}};
        Board initial = new Board(blocks);
        
        if (initial.isSolvable()) {
            Solver solver = new Solver(initial);
            System.out.println("Minimum number of moves = " + solver.moves());
            
            for (Board b : solver.soluction()) {
                System.out.println(b);
            }
        }
    }
}

File: Python/8puzzle Karen/solver.py
# solve.py

from board import Board
from stack import Stack, MinPQ
from grafico import grafico

class Solver(object):

	class _Node(object):

		def __init__(self, board, moves = 0, prev = None):
			self.board = board
			self.moves = moves
			self.prev  = prev

		def priority(self):
			return self.board.manhattan() + self.moves

		def __cmp__(self, other):
			return self.priority() - other.priority()

	def __init__(self, board):
		self.pq = MinPQ()
		self.goal = None

		self.pq.insert(self._Node(board))

		while True:
			node = self.pq.delMin()

			if node.board.goal():
				self.goal = node
				break
			for neighbor in node.board.neighbors():
				n = self._Node(neighbor, node.moves + 1, node)
				if not node.prev is None:
					if neighbor != node.board:
						self.pq.insert(n)
				else:
					self.pq.insert(n)

	def moves(self):
		return self.goal.moves if not self.goal is None else -1

	def solution(self):
		stck = Stack()
		node = self.goal
		while not node is None:
			stck.push(node.board)
			node = node.prev
		return stck

if __name__ == '__main__':

	#blocks  = [[0,1,3],[4,2,5],[7,8,6]]
	#blocks = [[6,0,5],[8,7,4],[3,2,1]]
	#blocks = [[8,6,7],[2,5,4],[1,3,0]]
	#blocks = [[2,3,4,8],[1,6,0,12],[5,10,7,11],[9,13,14,15]]
	#blocks = [[5,1,8],[2,7,3],[4,0,6]]
	blocks = [[1,2,3,4],[6,10,7,8],[5,0,11,12],[9,13,14,15]]
	initial = Board(blocks)  
	# check if puzzle is solvable; if so, solve it and output solution
	if initial.isSolvable():
		solver = Solver(initial)
		print "Minimum number of moves = %r" % solver.moves()
		for board in solver.solution():
			print board
			g=grafico()
			print(g.grafica(board))
		g.termino()	
		
	else:
		print "Unsolvable"